### ISSUE TITLE
Implement quantiles regression

### ISSUE THREAD

**Issue by mdymczyk** [h2oai/h2o4gpu#32]

**Comment by pseudotensor** [comment]
Add the below type of functionality into an xgboost wrapper:

http://www.bigdatarepublic.nl/regression-prediction-intervals-with-xgboost/
https://www.snip2code.com/Snippet/1651850/Customized-loss-function-for-quantile-re
http://codegists.com/code/xgboost-regression/

e.g.
```
import numpy as np
 
 
def xgb_quantile_eval(preds, dmatrix, quantile=0.2):
    """
    Customized evaluation function that equals
    to quantile regression loss.
 
    Quantile regression is regression that
    estimates a specified quantile of target's
    distribution conditional on given features.
 
    @type preds: numpy.ndarray
    @type dmatrix: xgboost.DMatrix
    @type quantile: float
    @return: float
    """
    labels = dmatrix.get_label()
    return ('q{}_loss'.format(quantile),
            np.nanmean((preds >= labels) * (1 - quantile) * (preds - labels) +
                       (preds < labels) * quantile * (labels - preds)))
 
 
def xgb_quantile_loss(preds, dmatrix, quantile=0.2):
    """
    Computes first-order derivative of quantile
    regression loss and a non-degenerate
    placeholder for second-order derivative.
 
    Placeholder is returned instead of zeros,
    because XGBoost requires non-zero
    second-order derivatives. See this page:
    https://github.com/dmlc/xgboost/issues/1825
    to see why it is possible to use this trick.
    However, be sure that hyperparameter named
    `max_delta_step` is small enough to satisfy:
    ```0.5 * max_delta_step <=
       min(quantile, 1 - quantile)```.
 
    @type preds: numpy.ndarray
    @type dmatrix: xgboost.DMatrix
    @type quantile: float
    @return: (numpy.ndarray, numpy.ndarray)
    """
    try:
        assert 0 <= quantile <= 1
    except AssertionError:
        raise ValueError("Quantile value must be float between 0 and 1.")
 
    labels = dmatrix.get_label()
    errors = preds - labels
 
    left_mask = errors < 0
    right_mask = errors > 0
 
    grad = -quantile * left_mask + (1 - quantile) * right_mask
    hess = np.ones_like(preds)
 
    return grad, hess
 
 
# Example of usage:
# bst = xgb.train(hyperparams, train, num_rounds,
#                 obj=xgb_quantile_loss, feval=xgb_quantile_eval)
```

How done in scikit:

http://scikit-learn.org/dev/auto_examples/ensemble/plot_gradient_boosting_quantile.html#example-ensemble-plot-gradient-boosting-quantile-py
https://github.com/scikit-learn/scikit-learn/pull/9978

How h2o-3 does it:

http://docs.h2o.ai/h2o/latest-stable/h2o-docs/data-science/algo-params/distribution.html

Example use on kaggle:

https://www.kaggle.com/c/allstate-claims-severity/discussion/26440

**Comment by navdeep-G** [comment]
Can be used in xgb glm

**Comment by pseudotensor** [comment]
I don't see why this was closed.

**Comment by pseudotensor** [comment]
We still manage xgboost and just because xgboost related doesn't mean we don't have work to do.

**Comment by navdeep-G** [comment]
Sure, sounds good. Sorry for the premature closing of the issue.
